### PR TITLE
Adding RangeDelete and CreateQueue methods for QueueV2 SQL Implementation

### PR DESCRIPTION
**Pull request by prathyushpv**
**What changed?**
Adding RangeDeleteMessages() and CreateQueue() methods for SQL version of persistence.QueueV2, which supersedes the persistence.Queue interface.

**Why?**
We plan on using this for the upcoming history task DLQ project.

**How did you test it?**
Unit tests. Code has 100% unit test coverage.

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

**Review by MichaelSnowden — CHANGES_REQUESTED** [review]
There's actually some off-by-one errors in the range delete logic. It's kind of hard to explain in plain English, and part of the issue is because of confusing code I wrote for the Cassandra version, so I took the liberty of just coding up the fix because it should be re-used in Cassandra as well: https://github.com/temporalio/temporal/tree/snowden/ppv-dlq-queueV2

I also added a test which verifies that the messages are actually deleted for SQL, similar to what we have in Cassandra.

File: common/persistence/cassandra/queue_v2_store.go [hunk 1]
```diff
@@ -533,17 +448,28 @@ func GetQueue(
...
+	return maxMessageID + 1, nil
+}
+
+func (s *queueV2Store) getMaxMessageID(ctx context.Context, queueType persistence.QueueV2Type, queueName string) (int64, bool, error) {
 	var maxMessageID int64
 
 	err := s.session.Query(TemplateGetMaxMessageIDQuery, queueType, queueName, 0).WithContext(ctx).Scan(&maxMessageID)
 	if err != nil {
 		if gocql.IsNotFoundError(err) {
 			// There are no messages in the queue, so the next message ID is the first message ID.
```
Comment (MichaelSnowden) [inline common/persistence/cassandra/queue_v2_store.go:469]: This comment no longer applies. We're returning false here, so the first return value is ignored.

File: common/persistence/queue_v2_util.go [hunk 2]
```diff
@@ -54,22 +56,92 @@ func GetNextPageTokenForQueueV2(result []QueueV2Message) []byte {
...
+	queue *persistencespb.Queue,
+) (*persistencespb.QueuePartition, error) {
+	// Currently, we only have one partition for each queue. However, that might change in the future. If a queue is
+	// created with more than 1 partition by a server on a future release, and then that server is downgraded, we
+	// will need to handle this case. Since all DLQ tasks are retried infinitely, we just return an error.
+	numPartitions := len(queue.Partitions)
+	if numPartitions != 1 {
+		return nil, serviceerror.NewInternal(
+			fmt.Sprintf(
+				"queue with type %v and queueName %v has %d partitions, but this implementation only supports"+
```
Comment (MichaelSnowden) [inline common/persistence/queue_v2_util.go:99]: nit: It's easier to search for where error messages occurred in the code base if error logs are structured like this <some static message>: <something formatted>, so I think we should change this to be something like "queue without single partition detected: ..."

File: common/persistence/queue_v2_util.go [hunk 3]
```diff
@@ -54,22 +56,92 @@ func GetNextPageTokenForQueueV2(result []QueueV2Message) []byte {
...
+}
+
+// GetDeleteRange returns the range of messages to delete, and a boolean indicating whether any messages should be deleted.
+func GetDeleteRange(request DeleteRequest) (DeleteRange, bool) {
+	lastID := request.LastIDToDeleteInclusive
+	newMinMessageID := lastID + 1
+	if lastID >= request.ExistingMessageRange.MaxMessageID {
+		// Never actually delete the last message
+		lastID = request.ExistingMessageRange.MaxMessageID - 1
+	}
```
Comment (MichaelSnowden) [inline common/persistence/queue_v2_util.go:135]: I think there's actually still a bug here. For example, if the messages are 0,1,2, and the requested last ID is 4, then we'll delete messages 0,1, and set min_message_id to 5. Then, message 2 will become a zombie message that can never be deleted, and the next message will have id 5, which is too high. Can you fix and also add a test for this case?

**Review by MichaelSnowden — CHANGES_REQUESTED** [review]

File: common/persistence/queue_v2_util.go (hunk 2 above)
Comment (prathyushpv) [reply common/persistence/queue_v2_util.go]: Done

File: common/persistence/cassandra/queue_v2_store.go (hunk 1 above)
Comment (prathyushpv) [reply common/persistence/cassandra/queue_v2_store.go]: Done

File: common/persistence/queue_v2_util.go (hunk 3 above)
Comment (prathyushpv) [reply common/persistence/queue_v2_util.go]: Good catch! The actual issue is when we read messages. After the deletion (with request last ID 4), if we enqueue a message, it will still have ID 3. Then if we try to read messages, we will never get this message. (Because min message id is 5).  So we will lose messages in the queue.  Fixed the issue. I have refactored GetDeleteRange a little bit to make it bit more concise.

File: common/persistence/queue_v2_util.go [hunk 4]
```diff
@@ -54,22 +56,87 @@ func GetNextPageTokenForQueueV2(result []QueueV2Message) []byte {
...
+	MaxMessageID int64
+}
+
+type DeleteRange struct {
+	InclusiveMessageRange
+	NewMinMessageID int64
+}
+
+// GetDeleteRange returns the range of messages to delete, and a boolean indicating whether any messages should be deleted.
+func GetDeleteRange(request DeleteRequest) (DeleteRange, bool) {
```
Comment (MichaelSnowden) [inline common/persistence/queue_v2_util.go:129]: This implementation looks **much** better. I was thinking of some test cases for this, and I realized that the boolean we return doesn't exactly mean there are messages to delete. We might just update min message ID without deleting any messages. For example:

```go
DeleteRequest{
  LastIDToDeleteInclusive: 1,
  ExistingMessageRange: InclusiveMessageRange{
    MinMessageID: 1,
    MaxMessageID: 1,
  },
}
```

should return:

```go
DeleteRange{
  InclusiveMessageRange: InclusiveMessageRange{
    MinMessageID: 1,
    MaxMessageID: 0,
  },
  NewMinMessageID: 2,
}, true
```

I think we should clarify in the comment that it returns true if the min message ID should be updated, something like this:

```go
// GetDeleteRange returns the range of messages to delete, and a boolean indicating whether there is any update to be
// made: meaning either we should delete messages, update the min message ID, or both.
```

Another thing. It's kind of hard to write integration tests for this function, but it's quite easy to unit test it. I asked ChatGPT to come up with a bunch, and then I edited them here: https://gist.github.com/MichaelSnowden/77c1e5fee5af92ec55f202d61c435ce3

I think we should add these to a unit test just for this file.

File: common/persistence/sql/queue_v2.go [hunk 5]
```diff
@@ -65,79 +63,103 @@ func NewQueueV2(db sqlplugin.DB,
...
+
+	_, err := q.getQueueMetadata(ctx, q.Db, request.QueueType, request.QueueName)
+	if err != nil {
+		return nil, err
+	}
 	tx, err := q.Db.BeginTx(ctx)
 	if err != nil {
-		return nil, serviceerror.NewUnavailable(fmt.Sprintf("EnqueueMessage failed. Failed to start transaction. Error: %v", err))
+		return nil, serviceerror.NewUnavailable(fmt.Sprintf(
+			"EnqueueMessage failed for queue with type: %v and name: %v. BeginTx operation failed. error: %v",
```
Comment (MichaelSnowden) [inline common/persistence/sql/queue_v2.go:81]: nit: capitalize Error

File: service/history/queues/queuestest/executable_dlq_test_suite.go [hunk 6]
```diff
@@ -75,7 +76,9 @@ func TestExecutable(t *testing.T, tqm persistence.HistoryTaskQueueManager) {
 			clusterName := "test-cluster-" + t.Name()
 
 			var executable queues.Executable = NewFakeExecutable(task, tc.err)
-			executable = queues.NewExecutableDLQ(executable, tqm, clock.NewEventTimeSource(), clusterName)
+			ts := clock.NewEventTimeSource()
+			ts.Update(time.Now())
```
Comment (MichaelSnowden) [inline service/history/queues/queuestest/executable_dlq_test_suite.go:80]: nit: Add a small comment about why we need this here

**Review by MichaelSnowden — APPROVED** [review]

File: common/persistence/queue_v2_util.go (hunk 4 above)
Comment (prathyushpv) [reply common/persistence/queue_v2_util.go:130]: That is a good idea! Added the tests.
Also updated the comment.

File: common/persistence/sql/queue_v2.go (hunk 5 above)
Comment (prathyushpv) [reply common/persistence/sql/queue_v2.go]: Done

File: service/history/queues/queuestest/executable_dlq_test_suite.go (hunk 6 above)
Comment (prathyushpv) [reply service/history/queues/queuestest/executable_dlq_test_suite.go:82]: Done
